### ISSUE TITLE
Shuffle triggers and automatic shuffling

### ISSUE THREAD

**Issue by ShawnieBoy** [magarena/magarena#576]
Currently shuffling is performed automatically whenever a card is put into a library from anywhere.

However, each card movement creates a shuffle, so any shuffle triggers will trigger for each card.

If `, then` was used as a link in regex similar to `~` shuffling could be performed as a separate action eg. `shuffle <your/target players'> library` This would also allow the scripting of cards which solely shuffle.

**Comment by melvinzhang** [comment]
Which spells/abilities cause multiple shuffle trigger to occur?

**Comment by ShawnieBoy** [comment]
Essentially any effect which moves multiple cards into a library:

Archangel's Light, Barishi - Graveyard to Library
Winds of Change - Hand to Library
Diminishing Returns, Time Reversal - Hand and Graveyard to Library
Rishadan Pawnshop, Unravel the Æther - Permanent to Library
Elixir of Immortality - Permanent and Graveyard to Library
Lich's Mirror, Sway of the Stars - Hand, Graveyard and Permanent to Library

**Comment by melvinzhang** [comment]
Thanks, needed some concrete examples for testing.

**Comment by melvinzhang** [comment]
What about examples of shuffle triggers?

**Comment by ShawnieBoy** [comment]
There isn't a trigger for 'when shuffle' because of this issue.

Cosi's Trickster
Psychic Surgery
Psychogenic Probe

**Comment by melvinzhang** [comment]
Moving to location OwnersLibrary is essentially a short hand for ShuffleIntoLibraryAction. If we don't want the shuffle, just move to either Top/BottomOfOwnersLibrary and perform the shuffle action manually. It should be easy to fix the multiple shuffle action bug. I'll look into it.

**Comment by melvinzhang** [comment]
Hmm, looking over MagicRuleEventAction, there are no effects that move cards into OwnersLibrary. All of the problematic cards are in groovy code.

**Comment by melvinzhang** [comment]
Unfortunately there is no way to fix all of these at once, it is a problem in the groovy implementation of some cards. Effect that move multiple cards into the library should first shift them to TopOfOwnersLibrary, then issue a single shuffle action at the end. If there are several cards, I think it makes sense to create a new Action specifically for this.

**Comment by melvinzhang** [comment]
@ShawnieBoy can you look into fixing each of the groovy code of the buggy cards? I'm having problems identifying the cards that need fixing.

**Comment by ShawnieBoy** [comment]
ShiftCardAction still causes shuffling. It uses MoveCardAction, which uses ShuffleIntoLibrary if the target is a library.

**Comment by melvinzhang** [comment]
Yes, shifting to location OwnersLibrary is essentially a short hand for ShuffleIntoLibraryAction.

**Comment by ShawnieBoy** [comment]
I'm not sure I understand - the cards aren't buggy yet as there's no actual trigger mechanism in place when shuffling the deck. If the trigger was placed in ShuffleIntoLibrary there would be multiple shuffles.

**Comment by melvinzhang** [comment]
The bug is shifting multiple cards into library with ShiftCardAction, it is meant for one cards. For multiple cards the correct way is to ShiftCardAction to TopOfOwnersLibrary followed by ShuffleLibraryAction.

**Comment by melvinzhang** [comment]
Yes, they are not causing any problems now, but if you want to add the shuffle trigger (which I assume is the purpose of this issue), they will cause problems because of their improper usage of ShiftCardAction.

**Comment by melvinzhang** [comment]
They don't have to fixed now (as you said, no issues when there are no shuffle trigger), just need to note that these groovy needs updating to use the correct implementation before shuffle triggers are introduced.

**Comment by ShawnieBoy** [comment]
I'm with you now, hehe.

I think adding a ShuffleAction to ShuffleIntoLibraryAction or removing it completely would help clarify things.

MoveCardAction would need to be edited to stop it from using ShuffleIntoLibraryAction

**Comment by ShawnieBoy** [comment]
Confusing myself - brain not in gear - To top! arrgh

**Comment by melvinzhang** [comment]
> MoveCardAction would need to be edited to stop it from using ShuffleIntoLibraryAction
Since this has been around for a long time, certain things depend on this. Off the top of my head, things like moving cards after they are cast uses MoveCardAction. Certain cases cause card which are cast/countered to go back to the library.

**Comment by melvinzhang** [comment]
Agree, ShuffleIntoLibraryAction should be removed, so there is only one place where shuffle occurs (ShuffleLibraryAction) and that is the place to call MagicGame.executeTrigger.

**Comment by ShawnieBoy** [comment]
For permanents going into the library - RemoveFromPlayAction would need to use ShiftCardAction to prevent auto-shuffling from MoveCardAction into library.

**Comment by ShawnieBoy** [comment]
I think Elixir of Immortality causes two shuffles - Once for the permanent and then for the graveyard. May even be two different players.

**Comment by melvinzhang** [comment]
For permanents, it shouldn't be ShiftCardAction (shift = remove + move), there is no card to remove for permanents. MoveCardAction is correct.

**Comment by melvinzhang** [comment]
Yes, Elixir of Immortality causes two shuffles. Should it be one? There could be a case where it is two though, as you alluded to, when the owner's are different.

**Comment by ShawnieBoy** [comment]
Well, currently it would have to be two due to the MoveCardAction - but there could be arguments for just one shuffle. I've had quick search for rulings, but not seen anything.

**Comment by melvinzhang** [comment]
The solution for RemoveFromPlay is similar to Shift, if there are multiple RemoveFromPlay to TopOfLib first, then do a single shuffle.

**Comment by ShawnieBoy** [comment]
Ahh yes, good thinking.

**Comment by melvinzhang** [comment]
Added a ShuffleCardsIntoLibraryAction in dbeba1d88554926a6b to make it easier to write in groovy, it does what I've explained here, first shift to top of lib, then do one shuffle at the end.

**Comment by ShawnieBoy** [comment]
arrgh, nearly finished the old stlye...

**Comment by ShawnieBoy** [comment]
Well, you'll be able to see in a bit that sometimes you need to shuffle from multiple locations at the same time.

**Comment by melvinzhang** [comment]
You can combine all the cards into a single MagicCardList then call ShuffleCardsIntoLibraryAction

**Comment by ShawnieBoy** [comment]
Would that include Permanents?

**Comment by melvinzhang** [comment]
Sorry, it doesn't work. Didn't think through it carefully, the new action only does for a single from location, so it is for the common case. For more complicate situations, need to do the shift to top + shuffle manually. It also doesn't work for permanents as that requires calling a different action (RemoveFromPlay). It is possible to generalize the ShuffleIntoLibrary action to work for more things. As a first cut, it is currently limited to cards from a same location.

**Comment by ShawnieBoy** [comment]
Let me finish what I've started then I'll look at what you've pushed

**Comment by ShawnieBoy** [comment]
I think I found all the shuffly-cards, and the ShuffleCardsIntoLibraryAction would help for most of them. Not sure if the shuffle check is needed as that would only apply to permanents - Other card-zones are player specific (Hand, Library, Graveyard).

Will have to look up some corner cases with whether shuffling an empty zone still causes a shuffle.

**Comment by ShawnieBoy** [comment]
Looks like it's still a shuffle.

**Comment by ShawnieBoy** [comment]
Actually, the check allows both players to be involved together which is handy. One list of both players' hands, graveyard and libraries all shuffling at once, but going to correct zones. :grinning:

**Comment by ShawnieBoy** [comment]
Instead of requiring a from, each card would be able to say where it is with a MagicCard.getLocation()

**Comment by melvinzhang** [comment]
Question: if permanent did not got the library due to shuffle effect, it is still necessary to shuffle? Eg "Shuffle target permanent into its owner's library" and "If this permanent would leave the battlefield, exile it instead of putting it anywhere else."

**Comment by melvinzhang** [comment]
My thinking is that there is no shuffle, because of the replacement of the "shuffle into library" with "go to exile".

**Comment by ShawnieBoy** [comment]
I would think more that the shuffle would happen, it would shuffle 0 cards - just like an empty graveyard.

**Comment by melvinzhang** [comment]
Looking through the groovy changes make me remember we had a RemoveAllFromPlayAction. That looks like the perfect place to do the remove from play to top + shuffle. So I've fixed that uses the same method as ShuffleCards in 47b42aff94d20b3a

**Comment by ShawnieBoy** [comment]
Tokens cause a shuffle, and they kinda work in a similar way to being exiled instead of anywhere else: Unravel the Æther on gatherer
>If the artifact or enchantment is a token, it will be shuffled into its owner’s library when Unravel the Æther resolves, then will cease to exist. For practical purposes, whatever item is representing the token should not actually be put into the library, though the library is still shuffled.

**Comment by melvinzhang** [comment]
Tokens is a bit different, they actually do go to the zone (graveyard, exile, library, etc). Then they disappear due to a state-based action. In this case, I'm referring to the case where the permanent never got to the zone due to a replacement effect.

**Comment by ShawnieBoy** [comment]
The new rules for Commanders should clear it up, I'll have a read - thinking more that the shuffle won't happen.

**Comment by ShawnieBoy** [comment]
>701.16c If an effect would cause a player to shuffle one or more specific objects into a library, that library is shuffled even if none of those objects are in the zone they’re expected to be in or an effect causes all of those objects to be moved to another zone or remain in their current zone.

**Comment by melvinzhang** [comment]
With the groovy updates, the multiple shuffling should be fixed. @ShawnieBoy does the changes resolve this issue adequately?

**Comment by melvinzhang** [comment]
40404d5124e8f5b8423a840649122e029935f753 adds shuffle your library as an action. @ShawnieBoy anything else left in this issue?

**Comment by ShawnieBoy** [comment]
Yes, I think everything's done except adding in a trigger.
